### PR TITLE
Configurations Enabling ICARUS NuMI Analysis

### PR DESCRIPTION
**Pull request by gputnam**
Gathers together a number of updates:

- Includes update to GENIE Systematics (from Jaesung)
- Includes build of Geant4Systematics
- Fix to MeVPrtl
- Updates to CAFMaker -- save hits from cathode crossers
- Updates to beam spill accounting (from Joseph+Jacob L)
- Updates to MCS (from Moon)

Depends on https://github.com/SBNSoftware/sbnanaobj/pull/104.

Marking as a draft, for now.

### CODE REVIEW

**Comment by jzennamo** [comment]
@gputnam it might help to base your branch off of my branch here:
https://github.com/SBNSoftware/sbncode/pull/394 

this has many similar changes as your PR

File: sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+  } // pm1sigma
+  else if (mode=="multisim"){
+    // multisim mode: parameter values sample within the given uncertainty for all parameters simultaneously
+    // Loop over universes j
+    for (unsigned j=0; j<fNsims; j++){
+      // In each multisim universe, loop through all parameters. For each parameter, generate a new random number from Nominal-Sigma to Nominal+Sigma.
+      std::map<std::string, double> tmp_vals;
+      for (size_t i_parset=0; i_parset<n_parsets; ++i_parset){
+        double r = fGaussRandom->fire(0.0,1.0);
+        tmp_vals[FitParNames.at(i_parset)] = FitParNominals.at(i_parset)+(FitParSigmas.at(i_parset)*r);
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx:213]: What is this supposed to be doing? I guess I just am not sure what `FitParSigmas.at(i_parset)` is supposed to be in this case

File: sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+
+
+    const simb::MCParticle& p = *mcparticles.at(i);
+    p_PDG = p.PdgCode();
+    int mcpID = p.TrackId();
+    std::string EndProcess  = p.EndProcess();
+
+    double mass = 0.;
+    if( TMath::Abs(p_PDG) == 211 ) mass = 139.57;
+    else if( p_PDG == 2212 ) mass = 938.28;
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx:278]: Doesn't `simb::MCParticle` have the `Mass()` is there a reason to not use that?

File: sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+      } // end loop over trajectory points
+
+      // Now find daughters of the MCP
+      std::vector<int> daughter_PDGs;
+      std::vector<int> daughter_IDs;
+      for( int i_mcp = 0; i_mcp < p.NumberDaughters(); i_mcp++ ){
+        int daughterID = p.Daughter(i_mcp);
+        for (auto test_mcp : mcparticles){
+          if (test_mcp->TrackId() == daughterID){
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx:332]: I thought `daughterID` was the location in the array of the `.Daughter(i_mcp)` and not the trackID:
https://internal.dunescience.org/doxygen/MCParticle_8cxx_source.html#l00112

File: sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+        double deltaX = ( trajpoint_X.at(istep) - trajpoint_X.at(istep-1) );
+        double deltaY = ( trajpoint_Y.at(istep) - trajpoint_Y.at(istep-1) );
+        double deltaZ = ( trajpoint_Z.at(istep) - trajpoint_Z.at(istep-1) );
+
+        double len = sqrt(
+          std::pow( deltaX, 2 )  +
+          std::pow( deltaY, 2 )  +
+          std::pow( deltaZ, 2 )
+        );
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx:375]:
```suggestion
        double len = std::hypot(deltaX, deltaY, deltaZ);
```

File: sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+        float w, el_w;
+
+        // I think this is the only bit that needs to change for different universes -- all the above is jut about the track, which doesn't change based on universe
+        ParMaker->SetNewVals(UniverseVals.at(j));
+        theReweighter->SetNewHists(ParMaker->GetFSHists());
+        theReweighter->SetNewElasticHists(ParMaker->GetElasticHist());
+        //Get the weight from the G4ReweightTraj
+        w = theReweighter->GetWeight( &theTraj );
+        // Total weight is the product of track weights in the event
+        weight[j] *= std::max((float)0.0, w);
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx:441]: should this be the max between 1 and w?

File: sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+        // Do the same for elastic weight (should be 1 unless set to non-nominal )
+        el_w = theReweighter->GetElasticWeight( &theTraj );
+        weight[j] *= std::max((float)0.0,el_w);
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/Calculators/Geant4/Geant4WeightCalc.cxx:445]: should this be the max between 1 and w?

File: sbncode/SBNEventWeight/jobs/geant4/eventweight_geant4_sbn.fcl
Comment (jzennamo) [inline sbncode/SBNEventWeight/jobs/geant4/eventweight_geant4_sbn.fcl:1]: Only comment is that maybe the random seeds should be a bit more....random ;)

File: sbncode/SBNEventWeight/jobs/geant4/piplus_reweight_parameters.fcl [hunk 7]
```diff
@@ -0,0 +1,66 @@
+BEGIN_PROLOG
+
+# From MicroBooNE, ubsim UBOONE_SUITE_v08_00_00_69
+
+PiPlusParameters: [
+  {
+    Name:       "fPiPlusReacLow"
+    Cut:        "reac"
+    Range:      [10., 200.]
+    Nominal:    1
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/jobs/geant4/piplus_reweight_parameters.fcl:10]:
```suggestion
    Nominal:    1.0
```

File: sbncode/SBNEventWeight/jobs/geant4/proton_reweight_parameters.fcl [hunk 8]
```diff
@@ -0,0 +1,23 @@
+BEGIN_PROLOG
+
+# From MicroBooNE, ubsim UBOONE_SUITE_v08_00_00_69
+
+ProtonParameters: [
+  {
+    Name:       "fProtonReac"
+    Cut:        "reac"
+    Range:      [10., 2005.]
+    Nominal:    1
```
Comment (jzennamo) [inline sbncode/SBNEventWeight/jobs/geant4/proton_reweight_parameters.fcl:10]:
```suggestion
    Nominal:    1.0
```

**Review by jzennamo — CHANGES_REQUESTED** [review]
Looks good, I might base your PR off my other PR

**Comment by gputnam** [comment]
This PR has now been superseded by a combination of other PRs, most recently: https://github.com/SBNSoftware/sbncode/pull/417, https://github.com/SBNSoftware/sbncode/pull/418/files, and https://github.com/SBNSoftware/sbncode/pull/419. Closing this in favor of those.

**Pull request closed without merging**